README.md
# NdT3 Discord Bot

This is my first Discord bot made using [Discord.js](https://github.com/hydrabolt/discord.js).

[![Codacy Badge](https://api.codacy.com/project/badge/Grade/a53801bdb675456da9c9e0f5794d26f5?branch=dev-branch)](https://www.codacy.com/app/NdT3Development/DiscordBot?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=NdT3Development/DiscordBot&amp;utm_campaign=Badge_Grade)

## It currently has these features:
  - Kicking users
  - Banning users
  - Ping command
  - Eval command
  - Purge
  - Say command (says what you tell it to say)
  - Stats
  - Stopping the bot using a command
  - Testing command (Testing, testing, 1 2 3)
  - Change settings command
  - Embed command (basic, using RichEmbed)
  - Deleted message logs
  - Invite link filter
  - Web Dashboard
  - Music Commands (Code from: https://github.com/iCrawl/Music-Bot)

## Upcoming Features

- Change of database

### Installation
First of all, clone this repository.

This bot requires [Node.js](https://nodejs.org/) v8+ (and npm) to run.

Also, please make a channel for mod logs in your Discord server.

Once you have that done, edit the `config.js.example` file and then rename the file to `config.js`. Make sure the callback URL specified is in the format of `{{http|https}}{{domain_name}}/callback` (Examples: `https://dashoard.bot-website.com/callback` OR `http://dashboard.bot-website.com/callback` OR `http://localhost:33445/callback`). This should be the public URL (proxied URL (by something like Nginx) is recommended).

![Bot Application](https://i.imgur.com/sechKvg.png)

### Example Configuration

#### config.js

```javascript
/* eslint-disable */
var config = { // NOTE: DO NOT LEAVE ANYTHING BLANK
	// ALL settings are strings. Do NOT just use true or false, use these in strings such as 'true' or 'false'. This is due to how some code works when changing the settings
	ownerID: '123456789012345678', // Your ID here
	token: 'Mxfxawx-token021kxxmkalpr-m', // Your bot token here
	status: 'dnd', // Bot status [online/idle/invisible/dnd]
	debug: 'false', // This is used to output some debug info if needed. The token will be in the console and other information could be in the console
	playingGame: '{{prefix}}help | {{guilds}} guilds | v{{version}}', // The game you want the bot to play. {{prefix}} is replaced with the default prefix below, {{guilds}} is replaced with the guild count and {{version}} is replaced with the bot version. Leave blank to disable
	purgeLogFormat: '\n Message ID: {{mID}} | Message Timestamp: {{mTS}} | Content: {{mC}} \n', // {{mID}}: Message ID; {{mTS}} Message Timestamp; {{mC}}: Message Content;
	eightBallResponses: ['Yes', 'No', 'Certainly', 'My sources say yes', 'Try again later', 'Without a doubt', 'Better not to tell you now'], // An array of responses for the 8ball command
	cleverbotToken: 'CC-3824abc', // API Token for CleverBot
  googleAPIToken: 'AI-241scsc', // Used for link shortener and music features. You need to have these APIs enabled.
  logTimeFormat: 'D MMM YYYY HH:mm:ss ZZ',
  musicEnabled: 'true',
	defaultSettings: {
		prefix: '!',
		modLogChannel: 'mod-log',
		modRole: 'Moderator',
		adminRole: 'Admin',
		welcomeChannel: 'general',
		welcomeMessage: 'Welcome {{user}}!',
		welcomeEnabled: 'false',
		inviteFilterEnabled: 'false',
		inviteWhitelist: ['discord-testers', 'discord-developers'], // This can be changed, these are just defaults as an example
		facepalms: 'false', // If enabled, the bot will reply with the facepalm emoji whenever a message contains 'facepalm'
		swearFilter: 'false',
		swearWords: ['damn'], // An array of swear words. These should be lowercase. (of course, I have not included much for certain reasons...)
		logDeletes: 'true',
		logNewMember: 'true',
		logMemberLeave: 'true',
		logCommandUsage: 'true',
		logPurge: 'true',
		sendHelp: 'channel' // Available options: channel, dm
	},
	dashboard: {
		enabled: 'true', // This setting controls whether the dashboard is enabled or not.
		oauthSecret: '0eFle4ArGsecret0sa', // The client secret from the Discord bot page
		secure: 'false', // HTTPS: 'true' for true, 'false' for false
		sessionSecret: '-crazyKeyboard-qwaszxerdfcvtyghbnuijkmopl', // Go crazy on the keyboard here, this is used as a session secret
		domain: 'dashboard.bot-website.com', // Domain name (with port if not running behind proxy running on port 80). Example: 'domain': 'dashboard.bot-website.com' OR 'domain': 'localhost:33445'
		port: '33445', // The port that it should run on
		invitePerm: '536079575',
		protectStats: 'false',
		borderedStats: 'false', // Controls whether stats in the dashboard should have a border or not
		legalTemplates: {
			contactEmail: 'support@dashboard.bot-website.com', // This email will be used in the legal page of the dashboard if someone needs to contact you for any reason regarding this page
			lastEdited: '18 November 2017' // Change this if you update the `TERMS.md` or `PRIVACY.md` files in `dashboard/public/`
		}
	}
};

module.exports = config;
```

In the config file above, the callbackURL would be set by the code to `https://dashboard.bot-website.com/callback`.

**It is recommended to run the dashboard with a proxy (like Nginx)**

#### Nginx Configuration

```
server {
    listen 80;

    server_name dashboard.bot-website.com;

    location / {
        proxy_pass http://127.0.0.1:33445;
    }
}
```

After that, you can install the dependencies and start the bot

**You can use the command line**
```sh
$ cd DiscordBot
$ npm install
$ npm start
```
**OR** start the bot using the `linux_run.sh` for Linux or `windows_run.bat` for Windows. _These files may be outdated._

**NOTE:** Running the bot with a process manager (like PM2) is recommended.

>This bot must be run on a Discord bot account. Do __NOT__ try to run this on a normal user account. This is against the Discord Terms of Service.

>Also, do __NOT__ play with the `eval` or `exex` command. You have been warned.

## Changes to the Code

You may change code if needed under the following conditions:

For the dashboard, you may change the theme, wording, design, links, etc. however I will not accept any bug reports coming from this. You will also agree to **not remove the copyright notice in the footer. You may add your name here, however, you must keep the original wording used**.

**ALL** copyright notices and credits **must** be kept as is, not edited in any way and not removed.

## Contributing

Want to contribute?

NdT3 Discord Bot is written in Discord.js. If you want to add a feature or work on the code, feel free make a pull request and your code might be accepted.

## Credits

- Bot based on AnIdiotsGuide's [example bot](https://github.com/An-Idiots-Guide/guidebot).
- Dashboard based on the one [here](https://idiots-dashboard.glitch.me/).
- Made using [Discord.js](https://github.com/hydrabolt/discord.js).
- The music part of the bot is based on [this bot](https://github.com/iCrawl/Music-Bot).

## Privacy:

Please read the `PRIVACY.md` file.

## License:

Please read the `LICENSE` file


CHANGELOG.md
# Changelog
**Changelog for NdT3's Discord bot**

This changelog was introduced in bot version 5.0.0-alpha.

__Not all changes will be posted here.__

---

## Version 8.3.0:

**New Commands!**

- Cleverbot
- Expand (expands short URLs)
- Level (Checks your level (works with point))
- Loadcommand (This allows you to load a new command without having to restart the bot)
- Lockdown (Locks a channel)
- Mute (Mutes a user (prevents them from talking))
- Permlevel (Check what permission level you have)
- Points (Check how many points you have)
- Reason (Allows a moderator to update a reason for a case)
- Remindme (Set reminders. Note: these will be reset when the bot restarts or the command is reloaded)
- Role (Allows moderators or above to give roles to members)
- Shorten (Allows you to shorten a long URL)
- Tags (Allows a moderator to store responses and a user to access these using `[prefix]tag tagname`)
- Weather (Now you can check the weather from Discord)

**New Music Commands!**

Music commands were added into the bot in this update:
- Np (Allows you to check what's playing)
- Pause (Pause the current song)
- Play (Plays a song)
- Queue (Check what's coming up)
- Resume (Continue a paused song)
- Skip (Allows you to skip a song)
- Stop (Stops the music, clears the queue, and leaves the channel)

*Please note that the music features are basic. I still need to add in a check for if there is no members in a voice channel that the music bot leaves that voice channel to save memory.*

**New Features!**

- Music Commands!
- Basic level system (You get points by sending messages)
- Case numbers

**Removed Feature**

- Removed the `blockConfigEval` option from the configuration file and removed this feature.

## Version 8.2.0-dev:

Big update, changelog coming later...

## Version 7.4.3:

- Removed some code that was not removed from a change in version 7.4.0
- Fixed up the `pin` command a bit
- Some CSS changes

## Version 7.4.2:

Fixed bad code in `purge` command

## Version 7.4.1:

Fixed the dashboard expand menu button being in the wrong position

## Version 7.4.0:

- Removed `dashboard.theme` configuration option
- Added a `Change Theme` button on the dashboard
- Added Cookie Consent as the above change uses cookies

## Version 7.3.1:

Bug fix

## Version 7.3.0:

New commands!
- `8ball`: Ask a yes/no question and the 8ball will give you an answer
- `pin`: Pins a message to the channel

New configuration options
- `eightBallResponses`: You can change the 8ball responses
- `dashboard.theme`: Choose between dark or light theme dashboard

- Included light theme CSS with the code under `dashboard/public` and renamed the dark theme CSS file

## Version 7.2.0:

- Added guild ID and guild owner ID to the `serverinfo` command
- Made a small change to the `dashboard.js` file which will now give the error code if there is an error starting the dashboard
- Included the CSS for the theme used in the dashboard (`https://bootswatch.com/darkly/`) as the Bootswatch hosted version of the CSS has URL changes which break the dashboard
- Minor bug fixes

## Version 7.1.0:

- Some changes to the dashboard to fix CSS issues
- A new configuration option: `dashboard.enabled`. This allows you to disable the dashboard
- Added some code to ensure that the bot code can only be used on a bot account
- Slight update to the PRIVACY.md file

## Version 7.0.0:

"A major update? Doesn't seem so major..." This is considered a major update due to it requiring you to reset the per guild settings before being able to use a new feature. Failure to do so will result in the feature not working or the bot even crashing. You can reset the per guild settings by running the bot command `[prefix]set reset` in a guild, or if that doesn't work, stopping the bot, then deleting the entire `data` directory, then starting the bot again.

Now enough of that, let's get into this short changelog update.

- Improved `userinfo` command. You can now check the info of another user!
- New feature in the `purge` command. LOGS! That's right, you can now have a log of the messages deleted by a purge. The log is in the form of a file attached to a message in the mod-log channel and also in the console. The purge command also brought the need for a new configuration option...
- New configuration option for per guild settings: `logPurge`. This option will allow you to set whether purges should have their messages logged.  _(As always, per guild configuration changes require a per guild setting reset. I will be working on a way to automatically add new settings with an `update` command)_
- New general configuration option: `purgeLogFormat`. This allows you to control the format of purge logs.

## Version 6.4.2:

Dashboard bug fixes

## Version 6.4.1:

Fixed a bug where commands may not run if logging is disabled

## Version 6.4.0:

Some dashboard changes and new dashboard config option

## Version 6.3.0:

Guild image in front of guild name in manage settings

## Version 6.2.0:

Added guild owner on dashboard

## Version 6.1.10:

Dashboard bug fix

## Version 6.1.9:

- Added a new commands field to `stats` embed.
- Fixed a problem in the dashboard

## Version 6.1.8:

Permissions bug fix

## Version 6.1.7:

Small change

## Version 6.1.6:

 Added more logging for DMs

## Version 6.1.4:

Should probably do local tests before pushing

## Version 6.1.3:

:eyes: No one saw that

## Version 6.1.3:

Fixed a bug where DM commands are not logged to console

## Version 6.1.2:

Fixed a `userinfo` bug

## Version 6.1.1:

`say` command disabled in DMs

## Version 6.1.0:

Getting a lot of work done on this bot.

- You can now DM the bot with certain commands! _(That was harder to implement than I thought it would be)_
- Moved the `createinvite` command to the moderation category

config.js
/* eslint-disable */
var config = { // NOTE: DO NOT LEAVE ANYTHING BLANK
    // ALL settings are strings. Do NOT just use true or false, use these in strings such as 'true' or 'false'. This is due to how some code works when changing the settings
    ownerID: '292065674338107393', // Your ID here
    token: 'NDIyNDc4NjQ5OTg0NTQ4ODgz.DfNjbA.cCd05Nr7kMlHZZLO0q7glmN9ecc', // Your bot token here
    status: 'dnd', // Bot status [online/idle/invisible/dnd]
    debug: 'false', // This is used to output some debug info if needed. The token will be in the console and other information could be in the console
    playingGame: 'none.', // The game you want the bot to play. {{prefix}} is replaced with the default prefix below, {{guilds}} is replaced with the guild count and {{version}} is replaced with the bot version. Leave blank to disable
    purgeLogFormat: '\n Message ID: {{mID}} | Message Timestamp: {{mTS}} | Content: {{mC}} \n', // {{mID}}: Message ID; {{mTS}} Message Timestamp; {{mC}}: Message Content;
    eightBallResponses: ['Yes', 'No', 'Certainly', 'My sources say yes', 'Try again later', 'Without a doubt', 'Better not to tell you now'], // An array of responses for the 8ball command
    cleverbotToken: 'CC-3824abc', // API Token for CleverBot
    googleAPIToken: 'AI-241scsc', // Used for link shortener and music features. You need to have these APIs enabled.
    logTimeFormat: 'LLLL',
    musicEnabled: 'true',
    defaultSettings: {
        prefix: 'm!',
        modLogChannel: 'mod-log',
        modRole: 'Moderador',
        adminRole: 'Administrador',
        welcomeChannel: 'general',
        welcomeMessage: 'Welcome {{user}}!',
        welcomeEnabled: 'false',
        inviteFilterEnabled: 'false',
        inviteWhitelist: ['discord-testers', 'discord-developers'], // This can be changed, these are just defaults as an example
        facepalms: 'false', // If enabled, the bot will reply with the facepalm emoji whenever a message contains 'facepalm'
        swearFilter: 'false',
        swearWords: ['damn'], // An array of swear words. These should be lowercase. (of course, I have not included much for certain reasons...)
        logDeletes: 'true',
        logNewMember: 'true',
        logMemberLeave: 'true',
        logCommandUsage: 'true',
        logPurge: 'true',
        sendHelp: 'channel' // Available options: channel, dm
    },
    dashboard: {
        enabled: 'true', // This setting controls whether the dashboard is enabled or not.
        oauthSecret: 'I5X0BsyGcNwJ78cvE7V7uNteBcFXrWBK', // The client secret from the Discord bot page
        secure: 'false', // HTTPS: 'true' for true, 'false' for false
        sessionSecret: 'none', // Go crazy on the keyboard here, this is used as a session secret
        domain: 'localhost:3030', // Domain name (with port if not running behind proxy running on port 80). Example: 'domain': 'dashboard.bot-website.com' OR 'domain': 'localhost:33445'
        port: '3030', // The port that it should run on
        invitePerm: '536079575',
        protectStats: 'false',
        borderedStats: 'false', // Controls whether stats in the dashboard should have a border or not
        legalTemplates: {
            contactEmail: 'support@dashboard.bot-website.com', // This email will be used in the legal page of the dashboard if someone needs to contact you for any reason regarding this page
            lastEdited: '18 November 2017' // Change this if you update the `TERMS.md` or `PRIVACY.md` files in `dashboard/public/`
        }
    }
};

module.exports = config;